### PR TITLE
feat: Intoduce Avoid throwing exceptions from unexpected locations check

### PR DESCRIPTION
**Pull request by ynsehoornenborg**
Added a new check to the existing `DocumentThrownExceptionsAnalyzer` which checks rule [8@102](https://tics.tiobe.com/viewerCS/?ID=105&CSTD=Rule) of the Philips C# Coding Standard

### CODE REVIEW

**Comment by ynsehoornenborg** [comment]
SonarCloud report a bug, which I feel is a bug in SonarCloud in this case, it doesn't seem to understand the (modern) `as` pattern syntax.

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidThrowingUnexpectedExceptionsAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	{
+		private const string LocationsTitle = @"Avoid throwing exceptions from unexpected locations";
+		private const string LocationsMessageFormat = @"Avoid throwing exceptions from {0}, as this location is unexpected.";
+		private const string LocationsDescription = @"Avoid throwing exceptions from unexpected locations, like Finalizers, Dispose, Static Constructors, etc...";
+		private const string Category = Categories.Documentation;
+
+		private static readonly DiagnosticDescriptor LocationsRule = new(Helper.ToDiagnosticId(DiagnosticIds.AvoidExceptionsFromUnexpectedLocations), LocationsTitle, LocationsMessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: false, description: LocationsDescription);
+
+		[SuppressMessage("", "IDE0090", Justification = "When type is removed the literal type cannot be deduced")]
+		private static readonly Dictionary<string, string> specialMethods = new Dictionary<string, string>
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidThrowingUnexpectedExceptionsAnalyzer.cs:29]: I noticed you don't use underscores for members. My teams have always done so, to easily distinguish from locals. (eg, when reviewing this, i spent a full 30 seconds trying to find the declaration of specialMethods in Analyze). The other arguably better approach is to use "this.".

**Review by bcollamore — APPROVED** [review]
Note Sonar

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidThrowingUnexpectedExceptionsAnalyzer.cs (hunk 1 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidThrowingUnexpectedExceptionsAnalyzer.cs:29]: Will do from now on !

**Pull request merged**